File: 백준/Silver/10799. 쇠막대기/쇠막대기.cc
#include <iostream>
#include <string>
#include <stack>
using namespace std;

int main() {
	stack<char> st;
	string input;
	int lazer = 0;
	
	
	cin >> input;
	for(int i = 0; i<input.size(); i++) {
		if(input.at(i) == '(') {
			st.push('(');
		}
		else if(input[i] == ')')
		{
			if(input[i-1] == '(') {
				st.pop();
				lazer += st.size();
			}
			else {st.pop();
				lazer++;
				
			}
		}
}
	
	cout << lazer;
	
	return 0;
}

File: 백준/Silver/10799. 쇠막대기/쇠막대기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String line = br.readLine();

        System.out.println(solution(line));
    }
    public static int solution(String input) {

        int level = 0;
        int answer = 0;

        Stack<Integer> st = new Stack<>();
        boolean check = false;//이전이 '(' 이면 1
        for (char c : input.toCharArray()) {
            if(c == '(') {
                level++;
                check = true;
            }

            else {
                if(check) {
                    answer += --level;
                    check = false;
                }
                else {
                    answer += 1;
                    level--;
                }
            }
        }

        return answer;
    }

}